cryptoconditions/types/zenroom.py

```diff
@@ -0,0 +1,303 @@
+                urlsafe_b64encode(self.script)),
+            'keys': base64_remove_padding(
+                urlsafe_b64encode(self.keys)),
+            # 'conf': base64_remove_padding(
+            #     urlsafe_b64encode(self.conf)),
+        }
+
+    # TODO Adapt according to outcomes of
+    # https://github.com/rfcs/crypto-conditions/issues/16
+    def parse_dict(self, data):
+        """
+        Generate fulfillment payload from a dict
+
+        Args:
+            data (dict): description of the fulfillment
+
+        Returns:
+            Fulfillment
+        """
+        if data.get('script'):
+            self.script = base58.b58decode(data['script'])
+        if data.get('keys'):
+            self.keys = base58.b58decode(data['keys'])
+
+    # TODO Adapt according to outcomes of
+    # https://github.com/rfcs/crypto-conditions/issues/16
+    def parse_json(self, data):
+        """
+        Generate fulfillment payload from a dict
+
+        Args:
+            data (dict): description of the fulfillment
+
+        Returns:
+            Fulfillment
+        """
+        self.script = urlsafe_b64decode(base64_add_padding(
+            data['script']))
+        self.keys = urlsafe_b64decode(base64_add_padding(
+            data['keys']))
+
+    def parse_asn1_dict_payload(self, data):
+        self.script = data['script']
+        self.keys = data['keys']
+
+    def validate(self, *, message):
+        """
+        Verify the signature of this Zenroom fulfillment.
+
+        The signature of this Zenroom fulfillment is verified against
+        the provided message and script.
+
+        Args:
+            message (str): Message to validate against.
+
+        Return:
+            boolean: Whether this fulfillment is valid.
+        """
+        message = json.loads(message)
+        data = {}
+        if 'data' in message['asset'].keys():
+            data['asset'] = message['asset']['data']
+        if self.data is not None:
+            data['output'] = self.data
+
+        # There could also be some data in the metadata,
+        # this is an output of the condition script which
+        # become an input for the fulfillment script
+        if message['metadata'] and 'data' in message['metadata']:
+            data['result'] = message['metadata']['data']
+
+        # We can put pulic keys either in the keys or the data of zenroom
+        data.update(self.keys)
+
+        result = ZenroomSha256.run_zenroom(self.script,
+                                           {},
+                                           data)
+
+        try:
+            result = json.loads(result.output)
+            # "Then print the string 'ok'" in zenroom produces a
+            # dictionary of array with the string 'ok'
+            # this is stored in result and compared against the content of
+            # the metadata
+            return result["output"][0] == message['metadata']['result']
+        except:
```

Comment:
could you catch the JSONDecodeError exception and potential exceptions in case message['metadata'] or message['metadata']['result'] does not exist? is it guaranteed that this exists?

Reply:
The code catches the exception, doesn'it? It is in a `try...except`, if `message['metadata']` or `message['metadata']['result']` the validation fails

Reply:
that's true, I just thought that it might be beneficial for the user/developer to get a better feedback instead of just having a functional result (validation failure). Especially in the initial phases, it will be good to get feedback like result output: could not be parsed, message did contain all expected fields, or similar. Just to ease debugging and investigation. I just read that the message format is actually not mandatory but you expect it in the validation. I think that's something we have to become clearer about or need to improve usability of the code

Reply:
Ok, validate could raise two exceptions:
- ZenroomException if the script could not be interpreted by Zenroom
- MalformedMessageException if the message doesn't include the keys metadata and result (that is if it doesn't contain the result of Zenroom)

The line `result = json.loads(result.output)` will never fail because it contains the result of Zenroom which is JSON (if it fails something really bad happened).